File: 223T/topsort.mjs
#!                                             /usr/bin/env node
import { readFileSync } from 'fs';;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
const WORDSIZ = 5;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
class fileReader {
    constructor( fileName) {
        this.buf = readFileSync( fileName, 'utf-8');;;;;;;;;;;;;
        this.j = 0;;;;;// buffer;position;;;;;;;;;;;;;;;;;;;;;;;
    }
    getNext() {
        if( this.buf[ this.j] === '\n') this.j++;// skip;;;;;;;;
        this.j += WORDSIZ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
        return +this.buf.substring( this.j - WORDSIZ, this.j);;; 
    }
}
var f = new fileReader( 'paper.dev');;;;;;;;;;;;;;;;;;;;;;;;;;;;
f.getNext();// skip(;dummy;record);;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
const n = f.getNext();// number;of;vertices;;;;;;;;;;;;;;;;;;;;;
function Link( sukc = 0) {
    this.sukc = sukc;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    this.next = null;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
}
function insert( u, v) {
    var node = new Link( v);;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    node.next = Top[ u].next;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    Top[ u].next = node;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
}
var Top = Array.from({ length: n + 1}, () => new Link());;;;;;;;
while( true) {
    const u = f.getNext();// u->v relation;;;;;;;;;;;;;;;;;;;;;;
    if(! u) break;// end of data;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    const v = f.getNext();;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    ++Top[ v].sukc;// predecessor COUNTER[ v]++;;;;;;;;;;;;;;;;;
    insert( u, v);// insert successive node v into Top[ u];;;;;;
}
var R = 0;// QUEUE rear pointer;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
for( let k = 1; k <= n; k++) {
    if(! Top[ k].sukc) R = Top[ R].sukc = k;;;;;;;;;;;;;;;;;;;;;
}
var F = Top[ 0].sukc;// QUEUE front pointer;;;;;;;;;;;;;;;;;;;;;
while( F) {
    console.log( F);;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    for( let node = Top[ F].next; node; node = node.next) {
        const k = node.sukc;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
        Top[ k].sukc--;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
        if(! Top[ k].sukc) R = Top[ R].sukc = k;;;;;;;;;;;;;;;;;
    }
    F = Top[ F].sukc;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
}
console.log( "'DONE");;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
////////////////////////////////////////////////////////////////
// log: 
////////////////////////////////////////////////////////////////

File: 223T/topsort.pl
#!                                                 /usr/bin/perl
################################################################
use warnings;
use strict;
use feature 'state'; # static variables
################################################################
my $filename = "./paper.dev";
#
open my $fh, "<", $filename or die "$filename: $!";
#
my $bufr = "";
#
while( my $line = <$fh>) { # Perl is the most funny language.
    chomp $line;           # 
    $bufr .= $line;        # 
}                          #
#
close $fh;
################################################################
my $MIXW = 5; # No. bytes in MIX words., cos the input file to
              # be consistent with the mixal program.
sub getnext {
    state $j = 0;
    $j += $MIXW;
    return 0 + substr $bufr, $j - $MIXW, $MIXW;
}
getnext(); # dummy record for padding, bcoz block size of mixal
# paper is 14, if not mistaken, words; and the input is in the
# form: u->v pairs,, for same reason the sentinel is 0, 0.
my $n = getnext();
print "No. vertices: $n\n";
####_###########################################################
my @cpre = (); # predecessor counters;
my @sucv = (); # successive lists
my $queue = \@cpre; # the queue trick
for( my $j = 0; $j <= $n; $j++) {
    push @cpre, 0;
    push @sucv, [];
}
sub insert_coin {
    my ($u, $v) = @_; # function arguments
    push @{$sucv[ $u]}, $v;
}
# build data structure #########################################
for( ;;) {
    my $u = getnext();
    my $v = getnext();
    if( $u == $v) {
        last;
    }
    $cpre[ $v]++; # increment ::v predecessors counter
    insert_coin( $u, $v); # push the tempo
}
sub dmp {
    for( my $j = 0; $j <= $n; $j++) {
        print "$j: ($cpre[ $j]) @{$sucv[ $j]}\n";
    }
}
dmp(); ########################################## fill the queue
my $r = 0; # rear pointer
for( my $j = 0; $j <= $n; $j++) {
    if( $cpre[ $j] == 0) {
        $r = $queue->[ $r] = $j;
    }
}
# Start poping #################################################
my $f = $queue->[ 0]; # front
print "topsort:";
while( $f != 0) {
    print " $f";
    # Decrement just pooped up ::f successors counters, and if
    # they become nil add them to the queue.
    for (@{$sucv[ $f]}) { # Here we have an example of three
        # different consecutive parenthesis.
        $cpre[ $_]--;
        if( $cpre[ $_] == 0) {
            $r = $queue->[ $r] = $_;
        }
    }
    $f = $queue->[ $f]; # moo forward
}
print "\n";
################################################################
